econml/selective_regularization.py

```diff
@@ -3,193 +3,98 @@
-        if sp.issparse(X):
-            X = X.toarray()
-        self.session.run(tf.global_variables_initializer())
-        for step in range(self._steps):
-            self.session.run(self.train, feed_dict={
-                self.X_reg: X[:, self._subset],
-                self.X_ureg: X[:, self._subset_c],
-                self.Y: y.reshape(-1, 1)
-            })
+        self._penalized_inds = np.delete(np.arange(X.shape[1]), self._unpenalized_inds)
+        X1 = X[:, self._unpenalized_inds]
+        X2 = X[:, self._penalized_inds]
+
+        X2_res = X2 - LinearRegression(fit_intercept=self._fit_intercept).fit(X1, X2).predict(X1)
+        y_res = y - LinearRegression(fit_intercept=self._fit_intercept).fit(X1, y).predict(X1)
+
+        self._penalized_model.fit(X2_res, y_res)
+
+        # The unpenalized model can't contain an intercept, because in the analysis above
+        # we rely on the fact that M(X beta) = (M X) beta, but M(X beta + c) is not the same
+        # as (M X) beta + c, so the learned coef and intercept will be wrong
+        intercept = self._penalized_model.predict(np.zeros_like(X2[0:1]))
+        if not np.allclose(intercept, 0):
+            raise AttributeError("The penalized model has a non-zero intercept; to fit an intercept "
+                                 "you should instead either set fit_intercept to True when initializing the "
+                                 "SelectiveRegression instance (for an unpenalized intercept) or "
+                                 "explicitly add a column of ones to the data being fit and include that "
+                                 "column in the penalized indices.")
+
+        # now regress X1 on y - X2 * beta2 to learn beta1
+        self._model_X1 = LinearRegression(fit_intercept=self._fit_intercept)
+        self._model_X1.fit(X1, y - self._penalized_model.predict(X2))
+
         return self
 
     def predict(self, X):
-        """Apply the model to a set of features to predict the outcomes."""
-        # TODO: any better way to deal with sparsity?
-        if sp.issparse(X):
-            X = X.toarray()
-        return self.session.run(self.Y_pred, feed_dict={
-            self.X_reg: X[:, self._subset],
-            self.X_ureg: X[:, self._subset_c]
-        })
+        X1 = X[:, self._unpenalized_inds]
+        X2 = X[:, self._penalized_inds]
+        return self._model_X1.predict(X1) + self._penalized_model.predict(X2)
 
     @property
     def coef_(self):
-        """Get the model coefficients."""
-        coef_reg = self.session.run(self.weights_reg.value())
-        coef_ureg = self.session.run(self.weights_ureg.value())
-        full_coef = np.zeros((coef_reg.shape[0] + coef_ureg.shape[0], coef_reg.shape[1]))
-        full_coef[self._subset, :] = coef_reg
-        full_coef[self._subset_c, :] = coef_ureg
-        return full_coef
+        coefs = np.empty(len(self._penalized_inds) + len(self._unpenalized_inds))
```

Comment:
This is wrong for multi-dimensional y. The following raises an error here.

```python
from econml.selective_regularization import SelectiveRegularization
from sklearn.linear_model import Lasso, ElasticNet, LassoCV, ElasticNetCV, MultiTaskElasticNetCV
import numpy as np
n = 1000
d = 100
X = np.random.normal(size=(n, d))
y = X[:, [0, 0]] + np.random.normal(size=(n, 2))
est = SelectiveRegularization([0, 1, 2], MultiTaskElasticNetCV(fit_intercept=False))
est.fit(X, y)
est.coef_
```

Reply:
e.g. the following would work:

```python
        if np.ndim(self._penalized_model.coef_) == 1:
            coefs = np.empty(len(self._penalized_inds) + len(self._unpenalized_inds))
            coefs[self._penalized_inds] = self._penalized_model.coef_
            coefs[self._unpenalized_inds] = self._model_X1.coef_
        else:
            coefs = np.empty((self._penalized_model.coef_.shape[0], len(
                self._penalized_inds) + len(self._unpenalized_inds)))
            coefs[:, self._penalized_inds] = self._penalized_model.coef_
            coefs[:, self._unpenalized_inds] = self._model_X1.coef_
```

Reply:
Fixed